master/buildbot/status/web/forms.py

```diff
@@ -45,55 +93,67 @@ def decodeFromURL(self, value, encoding):
-            cxt['slaves'] = sorted(cxt['slaves'], key=attrgetter('friendly_name'))
-
-            #Get branches
-            encoding = getRequestCharset(request)
-            branches = [ b.decode(encoding) for b in args.get("branch", []) if b ]
-            cxt['branches'] = branches
-
-            return_page = ""
-            if args.has_key("return_page"):
-                return_page = args['return_page']
-                if not isinstance(return_page, basestring):
-                    return_page = args['return_page'][0]
-
-            #Add scheduler info
-            buildForceContext(cxt, request, self.getBuildmaster(request), builder_status.getName())
-
-            #URL to force page with return param
-            url = args['builder_url'][0]
-            url_parts = list(urlparse(url))
-
-            if len(url_parts) > 0 and len(return_page) > 0:
-                return_page = "&returnpage={0}".format(return_page)
-            else:
-                return_page = "?returnpage={0}".format(return_page)
-            cxt['return_page'] = return_page
-
-            url_parts[2] += "/force"
-            url_parts[4] += return_page
-            force_url = urlunparse(url_parts)
-            cxt['force_url'] = force_url
-
-            authz = self.getAuthz(request)
-            cxt['is_admin'] = yield authz.getUserAttr(request, 'is_admin', 0)
+            buildMaster = self.getBuildmaster(request)
+
+            cxt['slaves'] = self._getSlaves(builder_status)
+            cxt['branches'] = self._getBranches(args, request)
+
+            buildForceContext(cxt, request, buildMaster, builder_status.getName())
+
+            url_parts = list(urlparse(args['builder_url'][0]))
+            cxt['return_page'] = self._getReturnPage(args, url_parts)
+            cxt['force_url'] = self._getForceUrl(cxt['return_page'], url_parts)
+
+            cxt['is_admin'] = yield self._getIsAdmin(request)
+            cxt['rt_update'] = args
+            request.args = args
+
+            template = request.site.buildbot_service.templates.get_template("force_build_dialog.html")
+            defer.returnValue(template.render(**cxt))
+
+        else:
+            page = ErrorPage(INTERNAL_SERVER_ERROR, "Missing parameters", "Not all parameters were given")
+            defer.returnValue(page.render(request))
+
+class RebuildDialogPage(BuildDialogPage, HtmlResource):
+    pageTitle = "Rebuild"
+
+    def decodeFromURL(self, value, encoding):
+        return urllib.unquote(value).decode(encoding)
+
+    @defer.inlineCallbacks
+    def content(self, request, cxt):
+        status = self.getStatus(request)
+        encoding = getRequestCharset(request)
+        args = self._decodeArgs(request, encoding)
+
+        # Get build info
+        buildNumber = int(self._getSingleArgument(args, encoding, "build_number"))
+        builderName = self._getSingleArgument(args, encoding, "builder_name")
+
+        if buildNumber != None and builderName != None:
```

Comment:
When `build_number` is incorrect integer (or it does not exist in `args`) then line 138 will raise `TypeError`

Reply:
👍